### PR TITLE
[SPARK-28124] [SS] SQS source for Structured Streaming

### PR DESCRIPTION
**Pull request by abhishekd0907**
## What changes were proposed in this pull request?

Using FileStreamSource to read files from a S3 bucket has problems both in terms of costs and latency:

- **Latency**: Listing all the files in S3 buckets every microbatch can be both slow and resource intensive.
- **Costs**: Making List API requests to S3 every microbatch can be costly.

The solution is to use Amazon Simple Queue Service (SQS) which lets you find new files written to S3 bucket without the need to list all the files every microbatch.

S3 buckets can be configured to send notification to an Amazon SQS Queue on Object Create / Object Delete events. For details see AWS documentation here [Configuring S3 Event Notifications](https://docs.aws.amazon.com/AmazonS3/latest/dev/NotificationHowTo.html) 

Spark can leverage this to find new files written to S3 bucket by reading notifications from SQS queue instead of listing files every microbatch.

This PR adds a new SQSSource which uses Amazon SQS queue to find new files every microbatch.

## Usage
`val inputDf = spark
.readStream
.format("s3-sqs")
.schema(schema)
.option("fileFormat", "json")
.option("sqsUrl", "https://QUEUE_URL")
.option("region", "us-east-1")
.load()`

## Implementation Details

We create a scheduled thread which runs asynchronously with the streaming query thread and periodically fetches messages from the SQS Queue. Key information related to file path & timestamp is extracted from the SQS messages and the new files are stored in a thread safe SQS file cache.

Streaming Query thread gets the files from SQS File Cache and filters out the new files. Based on the maxFilesPerTrigger condition, all or a part of the new files are added to the offset log and marked as processed in the SQS File Cache. The corresponding SQS messages for the processed files are deleted from the Amazon SQS Queue and the offset value is incremented and returned.

SQS Design (1)

## How was this patch tested?

Added new unit tests in SqsSourceOptionsSuite which test various SqsSourceOptions. Will add more tests after some initial feedback on design approach and functionality.

### CODE REVIEW

**Comment by gaborgsomogyi** [comment]
External connectors can be added outside of Spark (maybe Bahir).
Just a side note, defining only 3 negative tests looks... well... not really over-tested code.

File: external/s3-sqs/pom.xml [hunk 1]
```diff
@@ -0,0 +1,121 @@
...
+    </dependency>
+    <dependency>
+      <groupId>org.apache.spark</groupId>
+      <artifactId>spark-catalyst_${scala.binary.version}</artifactId>
+      <version>${project.version}</version>
+      <type>test-jar</type>
+      <scope>test</scope>
+    </dependency>
+    <dependency>
+      <groupId>com.amazonaws</groupId>
```
Comment (steveloughran) [inline external/s3-sqs/pom.xml]: The big aws shaded SDK which comes with hadoop 2.9++ hadoop-aws bundles this

File: external/s3-sqs/src/main/scala/org/apache/spark/sql/sqs/SqsClient.scala [hunk 2]
```diff
@@ -0,0 +1,267 @@
...
+    val messageList = try {
+      val receiveMessageRequest = new ReceiveMessageRequest()
+        .withQueueUrl(sqsUrl)
+        .withWaitTimeSeconds(sqsLongPollWaitTimeSeconds)
+      val messages = sqsClient.receiveMessage(receiveMessageRequest).getMessages.asScala
+      retriesOnFailure = 0
+      logDebug(s"successfully received ${messages.size} messages")
+      messages
+    } catch {
+      case ase: AmazonServiceException =>
```
Comment (steveloughran) [inline external/s3-sqs/src/main/scala/org/apache/spark/sql/sqs/SqsClient.scala]: I'm going to point you at the S3A logic to interpret AWS exceptions and remap
https://github.com/steveloughran/hadoop/blob/trunk/hadoop-tools/hadoop-aws/src/main/java/org/apache/hadoop/fs/s3a/S3AUtils.java#L179

And then retry policy we've evolved over time
https://github.com/steveloughran/hadoop/blob/trunk/hadoop-tools/hadoop-aws/src/main/java/org/apache/hadoop/fs/s3a/S3ARetryPolicy.java#L158

Each AWS service has its own failure modes which you get to learn over time.

File: external/s3-sqs/src/main/scala/org/apache/spark/sql/sqs/SqsClient.scala [hunk 3]
```diff
@@ -0,0 +1,267 @@
...
+    }
+  }
+
+  private def createSqsClient(): AmazonSQS = {
+    try {
+      val isClusterOnEc2Role = hadoopConf.getBoolean(
+        "fs.s3.isClusterOnEc2Role", false) || hadoopConf.getBoolean(
+        "fs.s3n.isClusterOnEc2Role", false) || sourceOptions.useInstanceProfileCredentials
+      if (!isClusterOnEc2Role) {
+        val accessKey = hadoopConf.getTrimmed("fs.s3n.awsAccessKeyId")
```
Comment (steveloughran) [inline external/s3-sqs/src/main/scala/org/apache/spark/sql/sqs/SqsClient.scala]: you should be using the fs.s3a values, given that that the S3n connector has been deleted from hadoop, though I'll add some more comments on that topic later

File: external/s3-sqs/src/main/scala/org/apache/spark/sql/sqs/SqsClient.scala [hunk 4]
```diff
@@ -0,0 +1,267 @@
...
+        AmazonSQSClientBuilder
+          .standard()
+          .withClientConfiguration(new ClientConfiguration().withMaxConnections(maxConnections))
+          .withCredentials(basicAwsCredentialsProvider)
+          .withRegion(region)
+          .build()
+      } else {
+        logInfo("Using the credentials attached to the instance")
+        val instanceProfileCredentialsProvider = new InstanceProfileCredentialsProviderWithRetries()
+        AmazonSQSClientBuilder
```
Comment (steveloughran) [inline external/s3-sqs/src/main/scala/org/apache/spark/sql/sqs/SqsClient.scala]: See: https://github.com/steveloughran/hadoop/blob/trunk/hadoop-tools/hadoop-aws/src/main/java/org/apache/hadoop/fs/s3a/S3AUtils.java#L1206

proxy support is essential; you can grab the keys from org.apache.hadoop.fs.s3a.Constants; things like timeout and buffer size a matter of preference

File: external/s3-sqs/src/main/scala/org/apache/spark/sql/sqs/SqsClient.scala [hunk 5]
```diff
@@ -0,0 +1,267 @@
...
+          .standard()
+          .withClientConfiguration(new ClientConfiguration().withMaxConnections(maxConnections))
+          .withCredentials(instanceProfileCredentialsProvider)
+          .build()
+      }
+    } catch {
+      case e: Exception =>
+        throw new SparkException(s"Error occured while creating Amazon SQS Client ${e.getMessage}")
```
Comment (steveloughran) [inline external/s3-sqs/src/main/scala/org/apache/spark/sql/sqs/SqsClient.scala]: needs to include the inner exception

File: external/s3-sqs/src/main/scala/org/apache/spark/sql/sqs/SqsClient.scala [hunk 6]
```diff
@@ -0,0 +1,267 @@
...
+        if (!batchResult.getFailed.isEmpty) {
+          batchResult.getFailed.asScala.foreach { entry =>
+            sqsClient.deleteMessage(
+              sqsUrl, requestEntries.get(entry.getId.toInt).getReceiptHandle)
+          }
+        }
+      }
+    } catch {
+      case e: Exception =>
+        logWarning(s"Unable to delete message from SQS ${e.getMessage}")
```
Comment (steveloughran) [inline external/s3-sqs/src/main/scala/org/apache/spark/sql/sqs/SqsClient.scala]: should log the exception. Also, some java exceptions (including NullPointerException) actually have a null .getMessage; the toString value is better

**Comment by steveloughran** [comment]
## General comments

it's good to see this; I've thought of adding this for a while.

1. if you pull in the `spark-hadoop-cloud` module then for Hadoop > 2.9 you get the full AWS SDK, shaded to avoid transitive classpath grief.
1. You get the `hadoop-aws` module s3a filesystem which should be the only ASF-open-source one to play with. Even on EMR it ends up on the CP, so you can use classes marked as @Public on it.

One thing that the S3A connector does is per-bucket config; when you instantiate the bucket it maps fs.s3a.bucket.BUCKET.* to fs.s3a.*. You can pick that up from the FS instance simply by using the configuration to bind to SQS from the filesystem, e.g

```
val fs=sourcePath.getFileSystem(conf)
val sqlsConf = fs.getConf();
```
for all other filesystems that will work, you just don't get the bucket override mechanism



## Auth

You can use the hadoop-aws credential providers declared as public/stable:
```
org.apache.hadoop.fs.s3a.SimpleAWSCredentialsProvider
org.apache.hadoop.fs.s3a.TemporaryAWSCredentialsProvider
```
*if you want to use some of the others we don't tag @Public, talk to me and I'll see whether we can mark them as such*

Plus the standard AWS ones
```
com.amazonaws.auth.EnvironmentVariableCredentialsProvider
com.amazonaws.auth.EC2ContainerCredentialsProviderWrapper
```
The latter one is recent; it can pick up credentials through containers as well as EC2 instances. 

If you are just going to read the hadoop config values, know that the recommended way to store them is actually in a Hadoop Credential Provider; `Configuration.getPassword()` retrieves them for either a credential source/file or falls back to an inline string value.

We have been doing some really fancy stuff in Hadoop 3.3 with Delegation Tokens, so that you can submit a spark job with local credentials and have a set of session/role credentials attached as a delegation token, then extracted at the far end. This allows you to submit Spark jobs with more rights than the deployed EC2 cluster and yet avoid passing in any long-lived secrets. 

We let DDB at these through `S3AFileSystem.shareCredentials()`; though that's a Hadoop 3.2+ API. I can backport it if you want to pick that up

Hope this helps
### Tests

Ignoring real-world integration tests (FWIW, I keep my downstream of spark in (https://github.com/hortonworks-spark/cloud-integration](https://github.com/hortonworks-spark/cloud-integration), you should be able to add some which test the JSON parsing of responses.

File: external/s3-sqs/src/main/scala/org/apache/spark/sql/sqs/SqsClient.scala (hunk 6 above)
Comment (abhishekd0907) [reply external/s3-sqs/src/main/scala/org/apache/spark/sql/sqs/SqsClient.scala]: will pick this change

File: external/s3-sqs/src/main/scala/org/apache/spark/sql/sqs/SqsClient.scala (hunk 5 above)
Comment (abhishekd0907) [reply external/s3-sqs/src/main/scala/org/apache/spark/sql/sqs/SqsClient.scala]: will pick this change

File: external/s3-sqs/src/main/scala/org/apache/spark/sql/sqs/SqsClient.scala (hunk 4 above)
Comment (abhishekd0907) [reply external/s3-sqs/src/main/scala/org/apache/spark/sql/sqs/SqsClient.scala]: yeah thanks for pointing out, will pick this

File: external/s3-sqs/src/main/scala/org/apache/spark/sql/sqs/SqsClient.scala (hunk 2 above)
Comment (abhishekd0907) [reply external/s3-sqs/src/main/scala/org/apache/spark/sql/sqs/SqsClient.scala]: I was following aws docs for covering various exception handling cases, they may not be very exhaustive though. Thanks for pointing to this code, it looks like a more sophisticated way to write the exception handling code.

**Comment by abhishekd0907** [comment]
@steveloughran I'm not sure if I should use hadoop 2.9. Spark parent already uses 2.7 as default hadoop version. Adding hadoop 2.9 as a dependency will increase the work of excluding / shading classes to avoid classpath conflicts & seems risky to me. Is it possible for you to backport the changes you want to hadoop 2.7 ?

**Comment by zsxwing** [comment]
> External connectors can be added outside of Spark (maybe Bahir).

+1. It can be a different project as well. We don't need to add this into Spark.

**Comment by steveloughran** [comment]
> @steveloughran I'm not sure if I should use hadoop 2.9. Spark parent already uses 2.7 as default hadoop version. Adding hadoop 2.9 as a dependency will increase the work of excluding / shading classes to avoid classpath conflicts & seems risky to me. Is it possible for you to backport the changes you want to hadoop 2.7 ?

2.7 is EOL, 2.8 is going to exactly one release as soon as we can find a volunteer and that's it tool. Neither are even qualified to work on Java 8 properly, so they need to be retired. It is time to move on. Sorry.


Bahir sounds like a home; I'll help review patches there. Make it all conditional on recent Hadoop 3.x+ and I can expose the relevant credential accessor there, maybe even think of a way of getting the AssumedRole delegation tokens to work with this too.

**Comment by steveloughran** [comment]
FWIW, thinking about this, an interesting way to do this (for S3A only), is for us to provide a way there to subscribe to SQS events on an s3a instance, e.g

```scala
val fs = bucketPath.getFileSystem(conf).asInstanceOf[S3AFileSystem]
val sqs = fs.subscribeToS3Events()
```

Good:
* makes it trivial to implement the subscriber
* lets s3a code handle all the existing proxy/connection/auth setup grief
* lets us right some tests for it in S3AFS ITest suites so you can be confident we don't break things

Con
* depends on a new release of Hadoop
* doesn't work with AWS EMRFS until they implement the same API.
* potentially less flexible for changes/fixes
```

I think for *me*, it'd work, but for you, not worth the complexity. And unless I can see a way to provide an event stream API which could be used across filesystems (e.g. implement an abfds and gcs equivalents; hdfs already has fadvise), you don't gain by having it added there. Sorry

**Comment by abhishekd0907** [comment]
As suggested by @steveloughran @gaborgsomogyi and @zsxwing, I'm closing this pull request. I've opened a pull request for this in Bahir. You guys can find it here and give your feedback:
[Faster S3 file Source for Structured Streaming with SQS](https://github.com/apache/bahir/pull/91)

**Comment by abhishekd0907** [comment]
> > @steveloughran I'm not sure if I should use hadoop 2.9. Spark parent already uses 2.7 as default hadoop version. Adding hadoop 2.9 as a dependency will increase the work of excluding / shading classes to avoid classpath conflicts & seems risky to me. Is it possible for you to backport the changes you want to hadoop 2.7 ?
> 
> 2.7 is EOL, 2.8 is going to exactly one release as soon as we can find a volunteer and that's it tool. Neither are even qualified to work on Java 8 properly, so they need to be retired. It is time to move on. Sorry.
> 
> Bahir sounds like a home; I'll help review patches there. Make it all conditional on recent Hadoop 3.x+ and I can expose the relevant credential accessor there, maybe even think of a way of getting the AssumedRole delegation tokens to work with this too.

@steveloughran Making it all conditional on Hadoop 3 will prevent Spark users who currently use default Hadoop version (2.7) from using it. We can make it all conditional on Hadoop 3 with Spark 3.0 onwards when Hadoop 3 becomes the default Hadoop Version for Spark. Right now, I would like my code to work for all spark users, even those who use spark with the default Hadoop version.

You can mention your other concerns on the [BAHIR pull request](https://github.com/apache/bahir/pull/91)

**Comment by gaborgsomogyi** [comment]
@abhishekd0907 thanks for pushing this forward. It will be a good addition.

**Pull request closed without merging**